src/sentry/api/endpoints/project_codeowners.py

```diff
@@ -1,11 +1,144 @@
+from sentry.ownership.grammar import (
+    parse_code_owners,
+    convert_codeowners_syntax,
+)
+
+from sentry.api.endpoints.project_ownership import ProjectOwnershipSerializer, ProjectOwnershipMixin
+
+logger = logging.getLogger(__name__)
+
+
+class ProjectCodeOwnerSerializer(CamelSnakeModelSerializer):
+    code_mapping_id = serializers.IntegerField(required=True)
+    raw = serializers.CharField(required=True)
+    organization_integration_id = serializers.IntegerField(required=False)
+
+    class Meta:
+        model = ProjectCodeOwners
+        fields = ["raw", "code_mapping_id", "organization_integration_id"]
+
+    def validate(self, attrs):
+        # If it already exists, set default attrs with existing values
+        if self.instance:
+            attrs = {
+                "raw": self.instance.raw,
+                "code_mapping_id": self.instance.repository_project_path_config,
+                **attrs,
+            }
+
+        if not attrs.get("raw", "").strip():
+            return attrs
+        external_association_err = ""
+        # Get list of team/user names from CODEOWNERS file
+        teamnames, usernames, emails = parse_code_owners(attrs["raw"])
+
+        # Check if there exists Sentry users with the emails listed in CODEOWNERS
+        users = User.objects.filter(email__in=emails)
+        users_emails = [user.email for user in users]
+        users_emails_diff = [email for email in emails if email not in users_emails]
+
+        if len(users_emails_diff):
+            external_association_err += f'The following emails do not have an user associated in Sentry: {", ".join(users_emails_diff)}.'
+
+        # Check if the usernames have an association
+        external_users = ExternalUser.objects.filter(external_name__in=usernames)
+
+        external_users_names = [user.external_name for user in external_users]
+        external_users_diff = [name for name in usernames if name not in external_users_names]
+
+        if len(external_users_diff):
+            external_association_err += f'The following usernames do not have an association in Sentry: {", ".join(external_users_diff)}.'
+
+        # Check if the team names have an association
+        external_teams = ExternalTeam.objects.filter(external_name__in=teamnames)
+
+        external_teams_names = [team.external_name for team in external_teams]
+        external_teams_diff = [name for name in teamnames if name not in external_teams_names]
 
+        if len(external_teams_diff):
+            external_association_err += f'The following team names do not have an association in Sentry: {", ".join(external_teams_diff)}.'
 
-class ProjectCodeOwnersEndpoint(ProjectEndpoint):
+        if len(external_association_err):
+            raise serializers.ValidationError({"raw": external_association_err})
+
+        # Convert CODEOWNERS into IssueOwner syntax
+        users_dict = {
+            user.external_name: user.organizationmember.user.email for user in external_users
```

Comment:
@NisanthanNanthakumar another place we probably need to use the `UserEmail` table

Reply:
@scefali the email in `auth_user` is the primary email in `UserEmail`. I think its ok to use this here bc we know the user exists.

Reply:
@NisanthanNanthakumar Good point, once we have the user we can use primary email here

Reply:
Should this organizationmember query have a condition for the current organization? This looks like it could cross organization boundaries.

Reply:
@markstory I updated the query for `external_teams` to check for the current organization here: https://github.com/getsentry/sentry/blob/7df73c982e1f8660ed680bd5b4e41ca18ca5063e/src/sentry/api/endpoints/project_codeowners.py#L64-L67

So in users_dict, all the users are in the organization of the project.